File: backend/backend/legacy/cs_chess_algorithms (pre-python-refactorization)/data_structures.cs
public enum ChessPiece {
    Pawn,
    Rook,
    Knight,
    Bishop,
    Queen,
    King
}

public enum DrawType {
    Stalemate,
    InsufficientMaterial,
    ThreefoldRepetition,
    FiftyMoveRule,
}

public class ChessFigure
{
    public ChessPiece type { get; set; }
    public bool isWhite { get; }

    public ChessFigure(ChessPiece type, bool isWhite)
    {
        this.type = type;
        this.isWhite = isWhite;
    }
    public ChessFigure(ChessFigure copy)
    {
        type = copy.type;
        isWhite = copy.isWhite;
    }
}

public struct moveInfo {
    //Coordinate notation retrieved information (ex. b5-c6, c4-c8, e7-e8-Q) <- Q means promoted to Queen
    public int sourceX { get; set; }
    public int sourceY { get; set; }
    public int targetX { get; set; }
    public int targetY { get; set; }
    public ChessPiece promotedTo { get; set; }

    //FEN retrieved information (ex. rnbqkbnr/pp1ppppp/8/2p5/4P3/8/PPPP1PPP/RNBQKBNR w KQkq c6 0 2)
    ChessFigure[,] chessBoard = new ChessFigure[8, 8];
    public bool whiteTurn { get; set; }
    public bool canWhiteSCastle { get; set; }
    public bool canWhiteLCastle { get; set; }
    public bool canBlackSCastle { get; set; }
    public bool canBlackLCastle { get; set; }
    public (int? x, int? y) enPassantTarget { get; set; }
    public int halfMoveClock { get; set; }
    public int moveNumber { get; set; }

    //Other retrieved information and useful data
    public ChessFigure sourcePiece { get; set; } //Retrieved by combining Coordinate + FEN board
    public ChessFigure targetPiece { get; set; } //Retrieved by combining Coordinate + FEN board
    public int distanceX { get; set; }
    public int distanceY { get; set; }
    public int directionX { get; set; }
    public int directionY { get; set; }
}

public struct NotationInfo {
    //Algebraic notation writing information
    public int sourceX { get; set; }
    public int sourceY { get; set; }
    public int targetX { get; set; }
    public int targetY { get; set; }
    public ChessPiece promotedTo { get; set; }
    public bool isLCastling { get; set; }
    public bool isSCastling { get; set; }
    public ChessPiece pieceType { get; set; }

    public bool isCapture { get; set; }
    public bool isCheck { get; set; }

    //FEN notation writing information
    ChessFigure[,] chessBoard = new ChessFigure[8, 8];
    public bool whiteTurn { get; set; }
    public bool canWhiteLCastle { get; set; }
    public bool canWhiteSCastle { get; set; }
    public bool canBlackLCastle { get; set; }
    public bool canBlackSCastle { get; set; }
    public (int? x, int? y) enPassantTarget { get; set; }
    public int halfMoveClock { get; set; }
    public int moveNumber { get; set; }

    //Game writing information (Also used for Algebraic notation)
    public bool isCheckmate { get; set; }
    public DrawType drawType { get; set; }
}

File: backend/backend/python_chess_algorithms/data_structures.py
from enum import Enum
from typing import Optional, Tuple, List

class ChessPiece(Enum):
    Pawn = 1
    Rook = 2
    Knight = 3
    Bishop = 4
    Queen = 5
    King = 6


class DrawType(Enum):
    Stalemate = 1
    InsufficientMaterial = 2
    ThreefoldRepetition = 3
    FiftyMoveRule = 4


class ChessFigure:
    def __init__(self, type: ChessPiece, is_white: bool):
        self.type = type
        self.is_white = is_white

    def __copy__(self):
        return ChessFigure(self.type, self.is_white)


class MoveInfo:
    def __init__(self):
        # Coordinate notation retrieved information (ex. 0543q)
        self.source_x: int = 0
        self.source_y: int = 0
        self.target_x: int = 0
        self.target_y: int = 0
        self.promoted_to: Optional[ChessPiece] = None

        # FEN retrieved information (ex. rnbqkbnr/pp1ppppp/8/2p5/4P3/8/PPPP1PPP/RNBQKBNR w KQkq c6 0 2)
        self.chess_board: List[List[Optional[ChessFigure]]] = [[None for _ in range(8)] for _ in range(8)]
        self.white_turn: bool = True
        self.can_white_s_castle: bool = False
        self.can_white_l_castle: bool = False
        self.can_black_s_castle: bool = False
        self.can_black_l_castle: bool = False
        self.en_passant_target: Optional[Tuple[Optional[int], Optional[int]]] = (None, None)
        self.half_move_clock: int = 0
        self.move_number: int = 1

        # Other retrieved information and useful data
        self.source_piece: Optional[ChessFigure] = None  # Retrieved by combining Coordinate + FEN board
        self.target_piece: Optional[ChessFigure] = None  # Retrieved by combining Coordinate + FEN board
        self.distance_x: int = 0
        self.distance_y: int = 0
        self.direction_x: int = 0
        self.direction_y: int = 0
        self.is_enpassant: bool = False


class NotationInfo:
    def __init__(self):
        # Algebraic and coordinate notation writing information
        self.source_x: int = 0
        self.source_y: int = 0
        self.target_x: int = 0
        self.target_y: int = 0
        self.promoted_to: Optional[ChessPiece] = None
        self.is_l_castling: bool = False
        self.is_s_castling: bool = False
        self.piece_type: Optional[ChessPiece] = None

        self.is_capture: bool = False
        self.is_enpassant: bool = False
        self.is_check: bool = False

        # FEN notation writing information
        self.chess_board: List[List[Optional[ChessFigure]]] = [[None for _ in range(8)] for _ in range(8)]
        self.white_turn: bool = True
        self.can_white_l_castle: bool = False
        self.can_white_s_castle: bool = False
        self.can_black_l_castle: bool = False
        self.can_black_s_castle: bool = False
        self.en_passant_target: Optional[Tuple[Optional[int], Optional[int]]] = (None, None)
        self.half_move_clock: int = 0
        self.move_number: int = 1

        # Game writing information (Also used for Algebraic notation)
        self.is_checkmate: bool = False
        self.draw_type: Optional[DrawType] = None